presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -749,12 +817,35 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
 
             for (int i = 0; i < node.getArguments().size(); i++) {
                 Expression expression = node.getArguments().get(i);
-                Type type = typeManager.getType(function.getArgumentTypes().get(i));
-                requireNonNull(type, format("Type %s not found", function.getArgumentTypes().get(i)));
-                if (node.isDistinct() && !type.isComparable()) {
-                    throw new SemanticException(TYPE_MISMATCH, node, "DISTINCT can only be applied to comparable types (actual: %s)", type);
+                Type expectedType = typeManager.getType(function.getArgumentTypes().get(i));
+                requireNonNull(expectedType, format("Type %s not found", function.getArgumentTypes().get(i)));
+                if (node.isDistinct() && !expectedType.isComparable()) {
+                    throw new SemanticException(TYPE_MISMATCH, node, "DISTINCT can only be applied to comparable types (actual: %s)", expectedType);
+                }
+                if (argumentTypes.get(i).hasDependency()) {
+                    FunctionType functionType = (FunctionType) expectedType;
+                    LambdaExpression lambdaExpression = (LambdaExpression) expression;
+                    List<LambdaArgumentDeclaration> lambdaArguments = lambdaExpression.getArguments();
+
+                    HashMap<String, LambdaArgumentDeclaration> nameToLambdaArgumentDeclarationMap = new HashMap<>();
+                    if (context.getContext().isInLambda()) {
+                        nameToLambdaArgumentDeclarationMap.putAll(context.getContext().getNameToLambdaArgumentDeclarationMap());
+                    }
+                    for (int j = 0; j < lambdaArguments.size(); j++) {
+                        LambdaArgumentDeclaration lambdaArgument = lambdaArguments.get(j);
+                        nameToLambdaArgumentDeclarationMap.put(lambdaArgument.getName(), lambdaArgument);
+                        expressionTypes.put(lambdaArgument, functionType.getArgumentTypes().get(j));
+                    }
+                    Type actualType = process(lambdaExpression.getBody(), new StackableAstVisitorContext<>(Context.inLambda(nameToLambdaArgumentDeclarationMap)));
+
+                    coerceType(lambdaExpression.getBody(), actualType, functionType.getReturnType(), format("Function %s argument %d", function, i));
+                    expressionTypes.put(lambdaExpression.getBody(), functionType.getReturnType());
+                    expressionTypes.put(lambdaExpression, functionType);
+                }
+                else {
+                    Type actualType = typeManager.getType(argumentTypes.get(i).getTypeSignature());
+                    coerceType(expression, actualType, expectedType, format("Function %s argument %d", function, i));
```

Comment:
This method `visitFunctionCall` has already about 200 lines of code. Can you please extract some methods with meaningful names. It would make this code much more readable.